src/sphinx_autodoc_typehints/__init__.py

```diff
@@ -334,13 +334,14 @@ def _resolve_type_guarded_imports(obj: Any) -> None:
                     for (_, part) in _TYPE_GUARD_IMPORT_RE.findall(module_code):
                         guarded_code = textwrap.dedent(part)
                         try:
-                            exec(guarded_code, obj.__globals__)
+                            with mock(app.config.autodoc_mock_imports):
```

Comment:
Let's pass in `app.config.autodoc_mock_imports` as argument and not the entire `app` object.